Content.Shared/_Crescent/Hardpoints/GridHardpointLimiterSystem.cs

```diff
@@ -0,0 +1,221 @@
+ //           Logger.Warning($"[HardpointLimiter] No tracker component found for grid {ev.gridUid}!");
+            return;
+        }
+
+        // Recalculate current count to be sure
+        var actualCount = GetCurrentHardpointCount(ev.gridUid);
+        comp.CurrentHardpoints = actualCount;
+
+//        Logger.Info($"[HardpointLimiter] Grid {ev.gridUid} - Current: {comp.CurrentHardpoints}, Max: {comp.MaxHardpoints}");
+
+        if (comp.CurrentHardpoints > comp.MaxHardpoints)
+        {
+ //           Logger.Warning($"[HardpointLimiter] Hardpoint limit exceeded on grid {ev.gridUid}! Current: {comp.CurrentHardpoints}, Max: {comp.MaxHardpoints}");
+        }
+    }
+
+    private void OnCannonDeanchored(HardpointCannonDeanchoredEvent ev)
+    {
+//        Logger.Info($"[HardpointLimiter] OnCannonDeanchored called for cannon {ev.CannonUid} on grid {ev.gridUid}");
+
+        if (!TryComp<ServerGridHardpointTrackerComponent>(ev.gridUid, out var comp))
+        {
+//            Logger.Warning($"[HardpointLimiter] No tracker component found for grid {ev.gridUid}!");
+            return;
+        }
+
+        // Recalculate current count to be sure
+        var actualCount = GetCurrentHardpointCount(ev.gridUid);
+        comp.CurrentHardpoints = actualCount;
+
+ //       Logger.Info($"[HardpointLimiter] Grid {ev.gridUid} after deanchor - Current: {comp.CurrentHardpoints}, Max: {comp.MaxHardpoints}");
+    }
+
+    public bool CanInstall(EntityUid gridUid)
+    {
+//        Logger.Info($"[HardpointLimiter] CanInstall called for grid {gridUid}");
+
+        if (!TryComp<ServerGridHardpointTrackerComponent>(gridUid, out var comp))
+        {
+//            Logger.Warning($"[HardpointLimiter] No tracker component found for grid {gridUid}, allowing installation");
+            return true;
+        }
+
+        var currentCount = GetCurrentHardpointCount(gridUid);
+//        Logger.Info($"[HardpointLimiter] Grid {gridUid} - Current count: {currentCount}, Component current: {comp.CurrentHardpoints}, Max: {comp.MaxHardpoints}");
+
+        // Update the component with actual count
+        comp.CurrentHardpoints = currentCount;
+
+        var canInstall = currentCount < comp.MaxHardpoints;
+//        Logger.Info($"[HardpointLimiter] CanInstall result: {canInstall} (current: {currentCount} < max: {comp.MaxHardpoints})");
+
+        return canInstall;
+    }
+
+    /// <summary>
+    /// Counts the number of anchored hardpoints on a grid
+    /// </summary>
+    private int GetCurrentHardpointCount(EntityUid gridUid)
+    {
+        var count = 0;
+        var query = AllEntityQuery<HardpointComponent, TransformComponent>();
+        while (query.MoveNext(out var uid, out _, out var xform))
+        {
+            if (xform.GridUid == gridUid && xform.Anchored)
+            {
+                count++;
+            }
+        }
+//        Logger.Info($"[HardpointLimiter] GetCurrentHardpointCount for grid {gridUid}: {count}");
+        return count;
+    }
```

Comment:
Hard no , this can be used to lag the server to hell. Use event-based counting where the canons add themselves to  the count instead of a full on server-wide query.